File: Week_06/213/house_robber_ii.go
package main

func rob(nums []int) int {
	if len(nums) == 0 {
		return 0
	} else if len(nums) == 1 {
		return nums[0]
	}
	return max(robHelper(nums[:len(nums)-1]), robHelper(nums[1:]))
}

func robHelper(nums []int) int{
	pre, cur := 0, 0
	for _, n := range nums {
		pre, cur = cur, max(cur, pre+n)
	}
	return cur
}

func max(x, y int) int {
	if x > y {
		return x
	}
	return y
}

File: Week_06/213/house_robber_ii.py
from typing import List


class Solution:
    def rob(self, nums: List[int]) -> int:
        def helper(nums: List[int]) -> int:
            pre = cur = 0
            for n in nums:
                pre, cur = cur, max(cur, pre + n)
            return cur

        if len(nums) == 0:
            return 0
        elif len(nums) == 1:
            return nums[0]
        return max(helper(nums[:-1]), helper(nums[1:]))
